From 7cbba4aa58a6eb4dce0806f48f3a2def591df303 Mon Sep 17 00:00:00 2001
From: Landen Chao <landen.chao@mediatek.com>
Date: Tue, 08 Dec 2020 18:31:48 +0800
Subject: [PATCH] eth: mediatek: disable phy auto polling

GMAC HW can polling the external phy link status and synchronize to MAC
via MDIO bus. Current implement uses linux kernel phylib to polling phy
status, so disable HW polling mechanism to ensure that the MDIO bus can
only be accessed by SW.

Test: mt7622 + mt7531 v13 rfb test pass

Change-Id: I8758ba450ccf04e305077bbe12c8731f7a995c40
CR-Id: WCNCR00181147
Signed-off-by: Landen Chao <landen.chao@mediatek.com>
---

diff --git a/drivers/net/ethernet/mediatek/mtk_eth_soc.c b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
index a3c14d0..c60d49f 100644
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -416,6 +416,7 @@
 static int mtk_mdio_init(struct mtk_eth *eth)
 {
 	struct device_node *mii_np;
+	u32 val;
 	int ret;
 
 	mii_np = of_get_child_by_name(eth->dev->of_node, "mdio-bus");
@@ -441,6 +442,10 @@
 	eth->mii_bus->priv = eth;
 	eth->mii_bus->parent = eth->dev;
 
+	/* Disable PHY auto-polling to avoid mdio access from HW */
+	val = mtk_r32(eth, MTK_PHY_PSC);
+	mtk_w32(eth, val & ~PHY_AP_EN, MTK_PHY_PSC);
+
 	snprintf(eth->mii_bus->id, MII_BUS_ID_SIZE, "%s", mii_np->name);
 	ret = of_mdiobus_register(eth->mii_bus, mii_np);
 
diff --git a/drivers/net/ethernet/mediatek/mtk_eth_soc.h b/drivers/net/ethernet/mediatek/mtk_eth_soc.h
index 1547e45..a52b185 100644
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.h
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.h
@@ -314,6 +314,10 @@
 #define RX_DMA_FPORT_SHIFT	19
 #define RX_DMA_FPORT_MASK	0x7
 
+/* PHY Polling and SMI Master Control */
+#define MTK_PHY_PSC		0x10000
+#define PHY_AP_EN		BIT(31)
+
 /* PHY Indirect Access Control registers */
 #define MTK_PHY_IAC		0x10004
 #define PHY_IAC_ACCESS		BIT(31)
